solutions/platinum/cf-1142B.mdx

```diff
@@ -0,0 +1,87 @@
+---
+id: cf-1142B
+source: Codeforces
+title: Lynyrd Skynyrd
+author: Justin Ji
+---
+
+<Spoiler title = "Hint 1">
+
+For each location in the array $a$, calculate the index of the next value
+needed in the permutation.
+
+</Spoiler>
+
+<Spoiler title = "Hint 2">
+
+For each left endpoint, calculate the minimum right endpoint for a range that
+has a suitable subsequence.
+
+</Spoiler>
+
+<Spoiler title = "Explanation">
+
+As mentioned in the hints, we calculate the index of the next value needed in
+the permutation. With this information, we can store an array $lift[x][y]$,
+that stores the minimum index to have $2^{x}$ values of our permutation covered,
+if we start at index $y$ in array $a$.
+
+After precalculating our binary lifts, we utilize binary lifting to calculate
+the minimum right endpoint for a cyclic shift of our permutation for each left
+endpoint. To calculate the best answer in the range $[l, r]$, we either use
+RMQ or store the prefix/suffix minimums of the answers.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(N\log{N})$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+int main() {
+	int n, m, q;
+	cin >> n >> m >> q;
+	vector<int> p(n);
+	vector<int> p_index(n);
+	for (int i = 0; i < n; i++) {
+		cin >> p[i];
+		p_index[--p[i]] = i;
+	}
+
+	vector<int> a(m);
+	for (int i = 0; i < m; i++) { cin >> a[i]; a[i]--; }
+
+	const int LOG = __lg(n) + 1;
```

Comment:
i mean you included bits, might as well use `std::log`
also since this isn't a hard constant perhaps rename the var to `log_n`

Reply:
fair enough